File: src/core/implementation/metadata/syncable_metadata.ts
import { Constants } from "../../abstraction/constants";
import { ISyncMetadata } from "../../abstraction/metadata/ISync_metadata";
import { SyncOperationEnum } from "../../abstraction/metadata/ISync_operation";
import { SyncZoneRestrictionEnum } from "../../abstraction/models/Sync_zone_restriction";

export class SyncMetadata implements ISyncMetadata{
    _id: string;
    timestamp: number;
    syncOperation: SyncOperationEnum;
    
  static create(metadata: SyncMetadata){
    let syncOperationMetadata = new SyncMetadata(metadata.type, metadata.version, metadata.syncZone);
    return syncOperationMetadata;
  }
    constructor(
        public type: string,
        public version: number,
        public syncZone:string) {}

      getSyncZone(){
          return this.syncZone;
      }

    computeSyncZone(syncZoneRestrictionType: SyncZoneRestrictionEnum){
      if(syncZoneRestrictionType == SyncZoneRestrictionEnum.restricted)
      return this.syncZone;
    else 
      return Constants.globalSyncZoneRestriction;
    }
}

File: fast_sync_dart_client/fast_sync_core/lib/src/implementation/metadata/syncable_metadata.dart
import 'package:fast_sync_client/src/absraction/constants.dart';
import 'package:fast_sync_client/src/absraction/metadata/isync_metadata.dart';
import 'package:fast_sync_client/src/absraction/metadata/isync_operation.dart';
import 'package:fast_sync_client/src/absraction/models/sync_zone_restriction.dart';

class SyncMetadata implements ISyncMetadata {
  @override
  final String id;

  @override
  final String syncZone;

  @override
  final String type;

  @override
  final num version;

  @override
  final num timestamp;

  @override
  final SyncOperationEnum syncOperation;

  const SyncMetadata(
      {required this.id,
      required this.syncZone,
      required this.type,
      required this.version,
      required this.timestamp,
      required this.syncOperation});

  @override
  String getSyncZone() {
    return syncZone;
  }

  @override
  String computeSyncZone(SyncZoneRestrictionEnum syncZoneRestrictionType) {
    if (syncZoneRestrictionType == SyncZoneRestrictionEnum.restricted) {
      return syncZone;
    } else {
      return Constants.globalSyncZoneRestriction;
    }
  }
}
